File: internal/server/projects.go
// projects.go — /api/v1/projects endpoints (gm-root.18).
//
// Two endpoints:
//
//   GET  /api/v1/projects          — enumerate projects under default_dir
//   POST /api/v1/projects/switch   — switch the active workspace
//
// Project discovery uses config.ListProjectsUnder (shared with the
// gm-root.17.4 cold-start gate so the scanning logic is not duplicated).
//
// "Switch workspace" in v1 is a lightweight server-side ack: it records
// which project the operator selected and returns the project entry so
// the SPA can update its route + topbar label. Full server-side workspace
// isolation (re-binding the WorkPlane adaptor) is out of scope for this
// bead — the SPA navigates to the selected project and the operator
// re-launches gemba serve against the new directory for isolated backends.

package server

import (
	"encoding/json"
	"net/http"
	"sync"

	"github.com/MikeBengtson/gemba/internal/config"
	"github.com/MikeBengtson/gemba/internal/server/httperr"
)

// projectEntry is one row in GET /api/v1/projects. Name is the directory
// basename; Path is the absolute disk path. Active is true for the
// project the operator most recently switched to (persisted in-process;
// resets on server restart).
type projectEntry struct {
	Name   string `json:"name"`
	Path   string `json:"path"`
	Active bool   `json:"active,omitempty"`
}

// projectsEnvelope is the response body of GET /api/v1/projects.
type projectsEnvelope struct {
	Projects []projectEntry `json:"projects"`
	Total    int            `json:"total"`
}

// switchProjectRequest is the request body of POST /api/v1/projects/switch.
// The caller must supply either Name or Path; Name is preferred and is
// matched against the project's directory basename.
type switchProjectRequest struct {
	// Name is the project directory basename (preferred).
	Name string `json:"name,omitempty"`
	// Path is the absolute project root. Honored when Name is empty.
	Path string `json:"path,omitempty"`
}

// switchProjectResponse is the response body of POST /api/v1/projects/switch.
type switchProjectResponse struct {
	Active projectEntry `json:"active"`
}

// activeProjectMu guards activeProjectName. The field is in-process
// state only — a server restart resets it to "" (no active project).
// gm-root.17.7 (start-planning handoff) and future beads that need the
// active-workspace field read it via Router.ActiveProject().
var (
	activeProjectMu   sync.RWMutex
	activeProjectName string
)

// listProjects handles GET /api/v1/projects.
//
// Shared with gm-root.17.4 (cold-start gate) via config.ListProjectsUnder.
// gm-root.17.7 (start-planning handoff) can reach the same project list
// through this endpoint.
func (r *Router) listProjects(w http.ResponseWriter, _ *http.Request) {
	ucfg, err := config.LoadUserConfig(r.cfg.ConfigPath)
	if err != nil {
		httperr.Write(w, http.StatusInternalServerError,
			"config_load_failed", "failed to load user config: "+err.Error())
		return
	}
	defaultDir, err := config.ResolveDefaultDir(ucfg)
	if err != nil {
		httperr.Write(w, http.StatusInternalServerError,
			"config_resolve_failed", "failed to resolve default_dir: "+err.Error())
		return
	}
	found, err := config.ListProjectsUnder(defaultDir)
	if err != nil {
		httperr.Write(w, http.StatusInternalServerError,
			"projects_scan_failed", "failed to enumerate projects: "+err.Error())
		return
	}

	activeProjectMu.RLock()
	active := activeProjectName
	activeProjectMu.RUnlock()

	out := make([]projectEntry, 0, len(found))
	for _, p := range found {
		out = append(out, projectEntry{
			Name:   p.Name,
			Path:   p.Path,
			Active: active != "" && p.Name == active,
		})
	}
	writeJSON(w, http.StatusOK, projectsEnvelope{
		Projects: out,
		Total:    len(out),
	})
}

// switchProject handles POST /api/v1/projects/switch.
//
// Records the selected project as active (in-process) and returns it so
// the SPA can update its route and topbar label. gm-root.17.7 (start-
// planning handoff) calls this endpoint after ratification to switch the
// workspace to the newly-created project.
func (r *Router) switchProject(w http.ResponseWriter, req *http.Request) {
	var body switchProjectRequest
	if err := json.NewDecoder(req.Body).Decode(&body); err != nil {
		httperr.Write(w, http.StatusBadRequest,
			"validation", "invalid request body: "+err.Error())
		return
	}
	if body.Name == "" && body.Path == "" {
		httperr.Write(w, http.StatusBadRequest,
			"validation", "name or path is required")
		return
	}

	ucfg, err := config.LoadUserConfig(r.cfg.ConfigPath)
	if err != nil {
		httperr.Write(w, http.StatusInternalServerError,
			"config_load_failed", "failed to load user config: "+err.Error())
		return
	}
	defaultDir, err := config.ResolveDefaultDir(ucfg)
	if err != nil {
		httperr.Write(w, http.StatusInternalServerError,
			"config_resolve_failed", "failed to resolve default_dir: "+err.Error())
		return
	}
	found, err := config.ListProjectsUnder(defaultDir)
	if err != nil {
		httperr.Write(w, http.StatusInternalServerError,
			"projects_scan_failed", "failed to enumerate projects: "+err.Error())
		return
	}

	// Find the requested project.
	var target *config.ProjectEntry
	for i := range found {
		p := &found[i]
		if body.Name != "" && p.Name == body.Name {
			target = p
			break
		}
		if body.Path != "" && p.Path == body.Path {
			target = p
			break
		}
	}
	if target == nil {
		httperr.Write(w, http.StatusNotFound,
			"project_not_found", "no project with that name or path was found")
		return
	}

	activeProjectMu.Lock()
	activeProjectName = target.Name
	activeProjectMu.Unlock()

	writeJSON(w, http.StatusOK, switchProjectResponse{
		Active: projectEntry{
			Name:   target.Name,
			Path:   target.Path,
			Active: true,
		},
	})
}

// ActiveProject returns the project name the operator most recently
// switched to, or an empty string if no switch has occurred this session.
// Exposed for gm-root.17.7 (start-planning handoff) so it can read the
// just-created project without re-scanning the disk.
func (r *Router) ActiveProject() string {
	activeProjectMu.RLock()
	defer activeProjectMu.RUnlock()
	return activeProjectName
}

File: web/src/api/projects.ts
// Typed data-access for /api/v1/projects (gm-root.18).
//
// Two endpoints:
//
//   GET  /api/v1/projects          — enumerate projects under default_dir
//   POST /api/v1/projects/switch   — switch the active workspace
//
// The active-workspace state is managed in ProjectPickerContext (see
// web/src/components/projectpicker/ProjectPickerContext.tsx), which
// holds the last-known active project and re-fetches the list on demand.
// gm-root.17.7 (start-planning handoff) calls switchProject after
// ratification; gm-root.17.4 (cold-start redirect) reads the project
// list via this module to decide whether /new is needed.

import { apiFetch } from './client';
import { CONFIRM_HEADER } from './workItems';
import { freshNonce } from './newproject';

// ProjectKind classifies a project's setup state (gm-root.17.14).
//
//   "complete"         — has both a beads DB and a .gemba/workspace.toml
//                        inside a git repo. Selecting it switches the
//                        active workspace.
//   "needs_workspace"  — beads DB exists but the directory has no
//                        .gemba/workspace.toml marker. The picker
//                        renders a "needs setup" badge and routes the
//                        click to BindDialog rather than switching.
//   "needs_repo"       — workspace marker exists but the directory is
//                        not inside a git worktree. Same picker
//                        treatment as needs_workspace; BindDialog
//                        offers `git init` here or "move into existing
//                        repo".
//
// Open string at the wire so future kinds don't break the SPA.
export type ProjectKind = 'complete' | 'needs_workspace' | 'needs_repo';

// ProjectEntry mirrors the server-side projectEntry shape from
// internal/server/projects.go. Name is the directory basename; Path
// is the absolute disk path; Active is true for the currently-selected
// project (in-process server state; resets on server restart). Kind
// classifies the project's setup state (gm-root.17.14) — entries
// whose kind != "complete" need to be bound via BindDialog before
// they can be selected as the active workspace.
export interface ProjectEntry {
  name: string;
  path: string;
  active?: boolean;
  kind: ProjectKind;
}

// ProjectsEnvelope is the response body of GET /api/v1/projects.
export interface ProjectsEnvelope {
  projects: ProjectEntry[];
  total: number;
}

// SwitchProjectRequest is the request body for POST /api/v1/projects/switch.
// Supply name (preferred, matched against the directory basename) or path
// (absolute disk path). Supplying both is fine — name is checked first.
export interface SwitchProjectRequest {
  name?: string;
  path?: string;
}

// SwitchProjectResponse is the response body of POST /api/v1/projects/switch.
export interface SwitchProjectResponse {
  active: ProjectEntry;
}

// listProjects fetches the full project list from GET /api/v1/projects.
// Returns an empty list (not an error) when no projects exist yet — the
// picker renders its empty state normally.
export async function listProjects(): Promise<ProjectsEnvelope> {
  return apiFetch<ProjectsEnvelope>('/v1/projects');
}

// switchProject posts to POST /api/v1/projects/switch and returns the
// newly-active project entry. Throws ApiError on network failure or if
// the project is not found (404).
//
// Shared endpoint for gm-root.17.7 (start-planning handoff after
// ratification) — that bead calls switchProject({ name }) and then
// navigates to the new project's board.
export async function switchProject(req: SwitchProjectRequest): Promise<SwitchProjectResponse> {
  return apiFetch<SwitchProjectResponse>('/v1/projects/switch', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(req),
  });
}

// AttachProjectRequest is the request body for POST /api/v1/projects/attach
// (gm-xwa8). Used by the configure-project modal to "adopt" an existing
// beads DB into a project on this machine. Exactly one of repo_path
// (existing git worktree) or create_at (new directory; gemba creates +
// `git init`s) must be set.
export interface AttachProjectRequest {
  beads_db: string;
  project_name: string;
  repo_path?: string;
  create_at?: string;
}

// AttachProjectResponse mirrors the server's projectEntry shape so the
// picker can re-render with the new entry without an extra round-trip.
export type AttachProjectResponse = ProjectEntry;

// attachProject posts to POST /api/v1/projects/attach. Nonce-gated so
// a double-submit can't double-write the workspace marker. Returns the
// newly-attached project entry on success.
export async function attachProject(
  req: AttachProjectRequest,
  opts: { nonce?: string } = {}
): Promise<AttachProjectResponse> {
  return apiFetch<AttachProjectResponse>('/v1/projects/attach', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      [CONFIRM_HEADER]: opts.nonce ?? freshNonce(),
    },
    body: JSON.stringify(req),
  });
}

// AdoptableDB is one row from GET /api/v1/projects/adoptable
// (gm-gmyl): a beads-shaped database visible on the configured Dolt
// server that does NOT correspond to any filesystem project. The
// picker renders these as one-click adopt candidates that pre-fill
// the configure-project modal.
export interface AdoptableDB {
  name: string;
  url: string;
}

// AdoptableEnvelope is the response body of /v1/projects/adoptable.
// `notice` is non-empty when the Dolt server is unreachable — the
// picker should still render normally; just hide the "Adoptable"
// section and surface the notice as a small inline hint.
export interface AdoptableEnvelope {
  dbs: AdoptableDB[];
  total: number;
  notice?: string;
}

// listAdoptableDBs fetches the adoptable list. Returns the envelope
// as-is so the caller can read both `dbs` and `notice`. Doesn't
// throw on Dolt-server-unreachable — that scenario surfaces as an
// empty list + a notice string.
export async function listAdoptableDBs(): Promise<AdoptableEnvelope> {
  return apiFetch<AdoptableEnvelope>('/v1/projects/adoptable');
}

// CloneProjectRequest is the request body for POST /v1/projects/clone
// (gm-e12.21.2). Both fields are required; the URL must be one of the
// supported schemes (http, https, ssh, scp-style git@host:org/repo).
export interface CloneProjectRequest {
  url: string;
  target_dir: string;
}

export interface CloneProjectResponse {
  repo_path: string;
  default_branch?: string;
  head_sha?: string;
}

// cloneProject posts to POST /v1/projects/clone. Nonce-gated so a
// double-submit can't double-clone. Throws ApiError on 4xx/5xx; the
// caller renders the error body's `message` inline.
export async function cloneProject(
  req: CloneProjectRequest,
  opts: { nonce?: string } = {}
): Promise<CloneProjectResponse> {
  return apiFetch<CloneProjectResponse>('/v1/projects/clone', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      [CONFIRM_HEADER]: opts.nonce ?? freshNonce(),
    },
    body: JSON.stringify(req),
  });
}

// BindProjectRequest is the request body for POST /api/v1/projects/bind
// (gm-root.17.14). The picker calls bindProject when an operator
// selects a project entry whose kind != "complete":
//
//   - mode: "create"   — `git init` a fresh repo at target_repo_path
//                        (which equals beads_db_path for the simple
//                        "create here" path) and write
//                        .gemba/workspace.toml.
//   - mode: "navigate" — the operator pasted an existing repo path;
//                        the server moves the beads DB into that repo
//                        and writes .gemba/workspace.toml.
//
// Both paths return the resolved entry with kind = "complete" so the
// SPA can switch the active workspace immediately on success.
export interface BindProjectRequest {
  beads_db_path: string;
  target_repo_path: string;
  mode: 'create' | 'navigate';
}

// BindProjectResponse mirrors the server's projectEntry shape so the
// picker can re-render with the now-complete entry without an extra
// round-trip.
export type BindProjectResponse = ProjectEntry;

// bindProject posts to POST /api/v1/projects/bind. Nonce-gated so a
// double-click on the confirm button can't double-write the workspace
// marker. Returns the now-complete project entry on success; throws
// ApiError with the server's error message on failure (target dir
// not a git repo, dest already exists, etc.) so the dialog can
// surface the message verbatim.
export async function bindProject(
  body: BindProjectRequest,
  opts: { nonce?: string } = {}
): Promise<BindProjectResponse> {
  return apiFetch<BindProjectResponse>('/v1/projects/bind', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      [CONFIRM_HEADER]: opts.nonce ?? freshNonce(),
    },
    body: JSON.stringify(body),
  });
}
